src/daemon/daemon.cpp

```diff
@@ -1982,50 +1937,52 @@ try // clang-format on
     fmt::memory_buffer errors;
     for (const auto& path_entry : request->target_paths())
     {
-        const auto name = path_entry.instance_name();
-        auto it = vm_instances.find(name);
-        if (it == vm_instances.end())
+        const auto& name = path_entry.instance_name();
+        const auto& target_path = path_entry.target_path();
+
+        if (vm_instances.find(name) == vm_instances.end())
         {
-            fmt::format_to(std::back_inserter(errors), "instance \"{}\" does not exist\n", name);
+            fmt::format_to(std::back_inserter(errors), "instance '{}' does not exist\n", name);
             continue;
         }
 
-        auto target_path = path_entry.target_path();
-        auto& mounts = vm_instance_specs[name].mounts;
+        auto& vm_spec_mounts = vm_instance_specs[name].mounts;
+        auto& vm_mounts = mounts[name];
+
+        auto do_unmount = [&](MountHandler& mount, const std::string& target) {
+            try
+            {
+                mount.stop();
+                vm_mounts.erase(target);
+                vm_spec_mounts.erase(target);
+            }
+            catch (const std::runtime_error& e)
+            {
+                fmt::format_to(std::back_inserter(errors), "failed to unmount \"{}\" from '{}': {}\n", target, name,
+                               e.what());
+            }
+        };
 
         // Empty target path indicates removing all mounts for the VM instance
         if (target_path.empty())
         {
-            for (const auto& mount_handler : config->mount_handlers)
+            for (auto it = vm_mounts.begin(); it != vm_mounts.end();)
             {
-                mount_handler.second->stop_all_mounts_for_instance(name);
+                // iterator must be advanced before used in order to prevent iterator invalidation caused by deleting
+                // from the iterated map
+                const auto& [target, mount] = *it++;
+                do_unmount(*mount, target);
             }
```

Comment:
I know this works, but it scares me a bit the fact that a function call can invalidate an iterator. What about unmounting everything and then emptying `vm_mounts`?

Reply:
I don't think that will work. Stopping and destructing a mount should happen in tandem. If we stop the mounts, but do not delete them from the map, if one of the mounts fails and throws from `stop`, the ones that already stopped will still be in the map when they souldn't.

Reply:
Why would it return an iterator to the first element? One alternative I see is for it to return an iterator to the next element after the deleted one, like `erase` does on a map. The best implementation for that approach I could come up with looks like this
```cpp
auto do_unmount = [&](auto mount_iter) {
    auto& [target, mount] = *mount_iter;
    try
    {
        mount->stop();
        vm_spec_mounts.erase(target);
        return vm_mounts.erase(mount_iter);
    }
    catch (const std::runtime_error& e)
    {
        fmt::format_to(std::back_inserter(errors), "failed to unmount \"{}\" from '{}': {}\n", target, name,
                       e.what());
        return ++mount_iter;
    }
}
// Empty target path indicates removing all mounts for the VM instance
if (target_path.empty())
{
    for (auto it = vm_mounts.begin(); it != vm_mounts.end(); it = do_unmount(it))
        ;
}
else
{
    if (auto it = vm_mounts.find(target_path); it != vm_mounts.end())
    {
        do_unmount(it);
    }
    else
    {
        fmt::format_to(std::back_inserter(errors), "path \"{}\" is not mounted in '{}'\n", target_path, name);
    }
}
```

Opinions?

Reply:
@ricab my concern was that the function `do_unmount()` invalidated the iterator but the caller had to take care of the invalid iterator. I proposed to address the invalidation in the same function, or something like that. But if you don't consider it's important, nevermind.

Reply:
Ahh ok, I see what you mean @luis4a0. `do_unmount` will need to change the iterator one way or another, so I guess the most important is to make that clear in the code. The lambda is already pretty localized and there is a clear comment, but I think receiving the iterator in `do_unmount` (as in Andrei's alternative code) makes more sense. 

I am not sure that incrementing it in the unmount lambda is the best. We don't always need the iterator any longer and unmount should be concerned with unmounting, not iterating. Perhaps renaming the param to `expiring_it` would make it clear enough that the client could not use the iterator any longer. Maybe even receiving by rvalue reference, even though copies would still be invalidated...

Reply:
Yeah, renaming to `expiring_it` is a good approach which self-documents the situation.

Reply:
One thing that I think should be taken into consideration is that this happens in other parts of the code (e.g. stopping/suspending an instance), but without the use of a lambda. The reason I used a lambda here is simple: the same code applies to both unmounting all mounts or one mount. But in other places, only all mounts are removed, so it doesn't make sense to have a lambda in those places, so the iterator is advanced inside the loop, with the same `auto& [target, mount] = *it++` trick. I think keeping it the same in all places makes it easier to understand.

Reply:
`*it++` is fine, my original worry is that it is used to take into account the invalidation done in another function. I agree it's better to handle the situation in the same way in all places, so maybe the renaming proposed by Ricardo is the best option.

Reply:
Okay, after all of this, I'm not entirely sure I understand what the final proposed solution is, so can you please spell it out for me? :sweat_smile:

Reply:
Alright, in today's sync, @luis4a0 clarified to me that the final solution is to leave the lambda as is and to better document the iterator invalidation through a better variable name and a comment.

Reply:
That and receiving the iterator directly in the lambda. That's the param I would call "expiring_it" (although naming the argument like that is fine too). As a bonus, that would avoid a subscript operation (computing hash and corresponding map node).